lib/cli/src/Cardano/CLI.hs

```diff
@@ -296,6 +331,202 @@ runCli = join . customExecParser preferences
+    }
+
+xPrvToHexText :: XPrv -> Text
+xPrvToHexText = T.pack . B8.unpack . hex . unXPrvStripPub
+
+unsafeHexTextToXPrv :: Text -> XPrv
+unsafeHexTextToXPrv t = either error Prelude.id $
+    xPrvFromStrippedPubXPrv
+    . unsafeFromHex
+    . B8.pack
+    . T.unpack
+    $ t
+
+-- | Map over the key type of a CliKeyScheme.
+--
+-- Can be used with e.g @xPrvToHexText@.
+mapKey
+    :: Functor m
+    => (key1 -> key2)
+    -> CliKeyScheme key1 m
+    -> CliKeyScheme key2 m
+mapKey f s = CliKeyScheme
+    { allowedWordLengths = allowedWordLengths s
+    , mnemonicToRootKey = fmap f . mnemonicToRootKey s
+    }
+
+eitherToIO :: Either String a -> IO a
+eitherToIO (Right a) = return a
+eitherToIO (Left e) = hPutStrLn stderr e >> exitFailure
+
+byronCliKeyScheme :: ByronWalletStyle -> CliKeyScheme XPrv (Either String)
+byronCliKeyScheme = \case
+    Random ->
+        let
+            proxy = Proxy @'Random
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap byronKeyFromSeed . seedFromMnemonic proxy)
+    Icarus ->
+        let
+            proxy = Proxy @'Icarus
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap icarusKeyFromSeed . seedFromMnemonic proxy)
+    Trezor ->
+        let
+            proxy = Proxy @'Trezor
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap hardwareKeyFromSeed . seedFromMnemonic proxy)
+    Ledger ->
+        let
+            proxy = Proxy @'Ledger
+        in
+            CliKeyScheme
+                (allowedLengths proxy)
+                (fmap hardwareKeyFromSeed . seedFromMnemonic proxy)
+
+  where
+    seedFromMnemonic
+        :: forall (s :: ByronWalletStyle).
+            (FromMnemonic (AllowedMnemonics s))
+        => Proxy s
+        -> [Text]
+        -> Either String SomeMnemonic
+    seedFromMnemonic _ =
+        left getFromMnemonicError . fromMnemonic @(AllowedMnemonics s)
+
+    allowedLengths
+        :: forall (s :: ByronWalletStyle). ( NatVals (AllowedMnemonics s))
+        => Proxy s
+        -> [Int]
+    allowedLengths _ =
+         (map fromIntegral (natVals $ Proxy @(AllowedMnemonics s)))
+
+    icarusKeyFromSeed :: SomeMnemonic -> XPrv
+    icarusKeyFromSeed = Icarus.getKey . flip Icarus.generateKeyFromSeed pass
+
+    byronKeyFromSeed :: SomeMnemonic -> XPrv
+    byronKeyFromSeed = Byron.getKey . flip Byron.generateKeyFromSeed pass
+
+    hardwareKeyFromSeed :: SomeMnemonic -> XPrv
+    hardwareKeyFromSeed = Icarus.getKey
+        . flip Icarus.generateKeyFromHardwareLedger pass
+
+    -- We don't use passwords to encrypt the keys here.
+    pass = mempty
+
+cmdRootKey :: Mod CommandFields (IO ())
+cmdRootKey =
+    command "root" $ info (helper <*> cmd) $ mempty
+        <> progDesc "Extract root xprv as hex\n\
+                    \ (64 bytes private key + 32 bytes chain code)"
+  where
+    cmd = do
```

Comment:
Why not follow the same pattern as other command-line handlers? All of them have separated the parsing of the cli argument from the execution of the command. 
This may sounds artificial, but it allows for better readability, better code-reuse (as the parsers can easily be re-used) and easier refactoring options (because the code is already decoupled). 

So: 

a) Declare a new data-type to carry the argument. (e.g. `TransactionListArgs`)
b) Declare your parsers with regards to the existing convention:
    - options are postfixed with "Option" (e.g. portOption, timeRangeOption, poolGapOption, paymentOption, randomPortOption, databaseOption, addressStateOption etc. ..) and declared under the relevant section.
    - arguments are postfixed with "Argument" (e.g. walletIdArgument, epochArgument, transactionIdArgument) 
    - Options are arguments are named after they actual name, camelCased, in the command-line to reduce discrepancy and make mapping between command-line arguments and their relevant parser easier. 

In this case, `parseKeyType` should really be called `typeOption` (although, I suggest to rename this in both the command line and here to perhaps something less ambiguous like `--wallet-style` or `--style` ?)

Reply:
Thanks for the explanations. I was too unfamiliar with everything here and didn't see the structure.

> although, I suggest to rename this in both the command line and here to perhaps something less ambiguous

Yeah,  `key root --wallet-style icarus` is probably clearer than `key root --type icarus` 🤷‍♂